### PR TITLE
Rework logic of TDS SSL Pre-login Handshake

### PR DESCRIPTION
**Pull request by thephantomthief**
This commit rewrites the logic of SSL prelogin handshake read. Previously, whatever segments we were reading from the SSL prelogin handshake packet, if the first byte started with TDS_PRELOGIN, then we assumed that the segment is actually the first segment of the SSL handshake packet from the client containing the TDS prelogin header.

This assumption is incorrect. It is possible that the segment being read by the SSL Handshake read method might not contain the TDS prelogin header but still has its first byte as TDS_PRELOGIN. There is no restriction on this for the encrypted handshake message.

We rewrite this logic to check for the TDS prelogin header only if we have fully read the entire previous SSL handshake packet from client or if this is actually the very first SSL packet from client (client hello).

Task: BABEL-4096
Signed-off-by: Sharu Goel <goelshar@amazon.com>

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by Deepesh125 — APPROVED** [review]

File: contrib/babelfishpg_tds/src/backend/tds/tdssecure.c [hunk 1]
```diff
@@ -69,72 +89,77 @@ SslRead(BIO * h, char *buf, int size)
...
-			if ((res = SslRead(h, buf, size)) <= 0)
-				return res;
+			header_left -= res;
 		}
-		else
-		{
-			int			tmp_res = 0;
-			int			i = TDS_PACKET_HEADER_SIZE;
 
-			for (i = TDS_PACKET_HEADER_SIZE; i < res; i++)
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tds/src/backend/tds/tdssecure.c:120]: Earlier we were ignoring header data, why is it OK now?

File: contrib/babelfishpg_tds/src/backend/tds/tdssecure.c [hunk 2]
```diff
@@ -69,72 +89,77 @@ SslRead(BIO * h, char *buf, int size)
...
+	Assert(pkt_bytes_read >= TDS_PACKET_HEADER_SIZE);
+
+	/*
+	 * If SSL packet expands the current TDS prelogin packet, we need to
+	 * read some data in next iteration.  For now, just read till end of the
+	 * packet.  The caller will try to read the remaining data in the next
+	 * iteration.
+	 * XXX: We've not really seen this scenario.
+	 */
+	if (pkt_bytes_read + size > pkt_data.header.length)
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tds/src/backend/tds/tdssecure.c:148]: does this mean that data is send as multiple pre-login packets and we are reading single packet at a time? How does this align with SSL handshake data?

File: contrib/babelfishpg_tds/src/backend/tds/tdssecure.c (hunk 1 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tds/src/backend/tds/tdssecure.c:120]: We are reading the header data only for fetching the packet length. Other than that we are still ignoring it as before and reading remaining data to account for the header bytes.

File: contrib/babelfishpg_tds/src/backend/tds/tdssecure.c (hunk 2 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tds/src/backend/tds/tdssecure.c:148]: As mentioned we haven't really seen this scenario. It is just an attempt to handle and log if we ever see such a scenario in SSL handshake. Previous logic was not addressing this at all.

Comment (surendravishnoi) [reply contrib/babelfishpg_tds/src/backend/tds/tdssecure.c:148]: we are handling it with assumption that additional data must belong to next packet. Have we made sure that it is always the case for test cases we have seen so far? SSL handshake will always be done using 1 or more pre-login packets where each packet has header in the start and followed by data tracked with length field.

Comment (kuntalghosh) [reply contrib/babelfishpg_tds/src/backend/tds/tdssecure.c:148]: As per the TDS protocol, the SSL packets will always be encapsulated in a TDS prelogin packet. It means two possible scenario:
1. One or more SSL packets are sent in a single TDS prelogin packet. We've seen this.
2. If the combined length SSL packets that should be sent together is more than TDS prelogin length, it'll be split in multiple TDS prelogin packets. We've not seen this scenario yet.

These possibilities are not specified in the protocol doc. But, we can assume it's a possibility if not explicitly mentioned in the doc. The above code tries to handle the second scenario. Another option is to throw an error as 'unhandled'. If we get an issue, we can fix the same.

**Review by surendravishnoi — APPROVED** [review]

**Pull request merged**